### PR TITLE
KernelSHAP / Lime Improvements

### PR DESCRIPTION
**Pull request by vivekmig**
* Adds support for generators as perturb function for Lime with corresponding tests
* Modifies KernelSHAP to sample based on categorical distributed on expected selected features and randomly sample vectors given expected number of selected features. This is theoretically equivalent to the previous approach of weighting randomly selected vectors, but this approach computationally scales better with larger numbers of features, since weights for larger numbers of features lead to arithmetic underflow.

### CODE REVIEW

**Review by miguelmartin75 — CHANGES_REQUESTED** [review]
First pass - mostly nits

File: captum/attr/_core/lime.py [hunk 1]
```diff
@@ -602,6 +622,31 @@ def default_perturb_func(original_inp, **kwargs):
...
+    if feature_mask is None:
+        feature_mask, num_interp_features = _construct_default_feature_mask(
+            formatted_inputs
+        )
+    else:
+        feature_mask = _format_input(feature_mask)
+        min_interp_features = int(
+            min(torch.min(single_inp).item() for single_inp in feature_mask)
+        )
+        if min_interp_features != 0:
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/lime.py:635]: Yeah, this would still technically work, since it would essentially create a feature vector of 100, but the extra features should be irrelevant, so results should match. But this would definitely be inefficient, so users should be careful of not leaving out feature values, which matches documentation. We could consider doing an extra step to remap non-consecutive IDs to a consecutive range, but it seems unnecessary if used correctly. I agree this warning is also not essential, and we could throw an exception, but since the current behavior is this way and it helps to avoid an extra step of just adjusting by a constant, it's probably fine to just leave it the same way for now.

File: captum/attr/_core/lime.py [hunk 2]
```diff
@@ -72,7 +73,7 @@ def __init__(
         forward_func: Callable,
         interpretable_model: Model,
         similarity_func: Callable,
-        perturb_func: Callable,
+        perturb_func: Union[Callable],
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/lime.py:76]: Good catch, thanks! Forgot to revert this change.

File: captum/attr/_core/kernel_shap.py [hunk 3]
```diff
@@ -29,22 +33,41 @@ def kernel_shap_similarity_kernel(
...
-    else:
+    if num_selected_features == 0 or num_selected_features == num_features:
         # weight should be theoretically infinite when denom = 0
         # enforcing that trained linear model must satisfy
         # end-point criteria. In practice, it is sufficient to
         # make this weight substantially larger so setting this
-        # weight to 100 (all other weights are < 1).
-        similarities = 100.0
+        # weight to 1000000 (all other weights are 1).
+        similarities = 1000000.0
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/kernel_shap.py:42]: Yes, it's a good point that this could need to be customized. For now, to avoid having too many parameters, I can make this an instance variable that advanced users can override on the object after creation, but we can make it a parameter later if necessary.

**Comment by vivekmig** [comment]
Thanks for the review @miguelmartin75 ! Addressed comments.

File: captum/attr/_core/lime.py [hunk 4]
```diff
@@ -481,7 +501,7 @@ def attribute(
             dataset = TensorDataset(
                 combined_interp_inps, combined_outputs, combined_sim
             )
-            self.interpretable_model.fit(DataLoader(dataset, batch_size=n_samples))
+            self.interpretable_model.fit(DataLoader(dataset, batch_size=batch_count))
```
Comment (NarineK) [inline captum/attr/_core/lime.py:504]: nit: Wouldn't `n_samples` still work here even in case of `perturb_generator` ?

File: tests/attr/test_lime.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env python3
 
 import unittest
-from typing import Any, Callable, Tuple, Union
+from typing import Any, Callable, Generator, Tuple, Union
```
Comment (NarineK) [inline tests/attr/test_lime.py:4]: A question about `alt_perturb_func`: in case of `default_perturb_func` we return only the probs but in alternative one it looks like we return masked values.  I might be missing something. I was curious.

File: tests/attr/test_kernel_shap.py [hunk 6]
```diff
@@ -197,6 +197,7 @@ def test_multi_input_batch_kernel_shap(self) -> None:
             expected,
             additional_input=(1,),
             feature_mask=(mask1, mask2, mask3),
+            n_perturb_samples=300,
```
Comment (NarineK) [inline tests/attr/test_kernel_shap.py:200]: nit: Why did we set, `n_perturb_samples` here ?

File: captum/attr/_core/kernel_shap.py [hunk 7]
```diff
@@ -294,8 +267,15 @@ def attribute(  # type: ignore
...
         """
-        return Lime.attribute.__wrapped__(
-            self,
+        formatted_inputs, baselines = _format_input_baseline(inputs, baselines)
+        feature_mask, num_interp_features = construct_feature_mask(
+            feature_mask, formatted_inputs
+        )
+        num_features_list = torch.arange(num_interp_features, dtype=torch.float)
+        denom = num_features_list * (num_interp_features - num_features_list)
+        probs = (num_interp_features - 1) / denom
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/kernel_shap.py:276]: Do we have a citation for this trick ?

File: captum/attr/_core/kernel_shap.py [hunk 8]
```diff
@@ -304,4 +284,48 @@ def attribute(  # type: ignore
...
+        if num_selected_features == 0 or num_selected_features == num_features:
+            # weight should be theoretically infinite when denom = 0
+            # enforcing that trained linear model must satisfy
+            # end-point criteria. In practice, it is sufficient to
+            # make this weight substantially larger so setting this
+            # weight to 1000000 (all other weights are 1).
+            similarities = self.inf_weight
+        else:
+            similarities = 1.0
+        return torch.tensor([similarities])
```
Comment (NarineK) [inline captum/attr/_core/kernel_shap.py:307]: This seem to become somewhat dummy in this case. It will always return one or the other constant.

**Review by NarineK — COMMENTED** [review]
Thank you for working on this PR, @vivekmig!
I left couple nits. I think that it would be good to describe this trick a little bit in the code since the original approach in the paper is a bit different in terms of the kernel similarity function.

File: captum/attr/_core/lime.py [hunk 9]
```diff
@@ -411,8 +417,22 @@ def attribute(
...
             for _ in range(n_samples):
-                curr_sample = self.perturb_func(inputs, **kwargs)
+                if perturb_generator:
+                    try:
+                        curr_sample = next(perturb_generator)
+                    except StopIteration:
+                        warnings.warn(
+                            "Generator completed prior to given n_samples iterations!"
+                        )
+                        break
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/lime.py:432]: Yes, I think this is needed since curr_sample wouldn't obtain a new sample for the remaining code in the loop. The alternative would be to raise an exception here rather than a warning, but it seems ok to just train with all samples returned from the generator, even if this is less than n_samples.

File: captum/attr/_core/lime.py (hunk 4 above)
Comment (vivekmig) [reply captum/attr/_core/lime.py:504]: The edge case is the same as above, essentially a provided generator which provides fewer than n_samples. A warning in this case (rather than an error) requires this change.

File: captum/attr/_core/kernel_shap.py [hunk 10]
```diff
@@ -304,4 +284,48 @@ def attribute(  # type: ignore
...
+
+    def kernel_shap_perturb_generator(
+        self, original_inp: Union[Tensor, Tuple[Tensor, ...]], **kwargs
+    ) -> Generator[Tensor, None, None]:
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/kernel_shap.py:311]: Sure, the generator is only yielding tensors, I can add a comment to make it clearer. The syntax for generator type hinting is defined here ( https://www.python.org/dev/peps/pep-0484/#annotating-generator-functions-and-coroutines ), the first type is the yield type, which should be Tensor. The second type is the send (input) type and the third type is a return type, which is provided only at StopIteration, in this case, I believe both should be None.

File: tests/attr/test_lime.py (hunk 5 above)
Comment (vivekmig) [reply tests/attr/test_lime.py:4]: This is based on the perturb_interpretable_space option, for the default, it is set to True, while for alt_perturb_func, we use perturb_interpretable_space=False.

File: captum/attr/_core/kernel_shap.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/kernel_shap.py:276]: Yes, denom is zero at index 0, so the division leads to an inf value, which is set to 0 in the next line. Yes, num_features_list would cover all possibilities (0 to num features). The original implementation does something similar, the paper is already cited, but I can link the code as well.

File: captum/attr/_core/kernel_shap.py [hunk 11]
```diff
@@ -294,8 +267,15 @@ def attribute(  # type: ignore
...
-        return Lime.attribute.__wrapped__(
-            self,
+        formatted_inputs, baselines = _format_input_baseline(inputs, baselines)
+        feature_mask, num_interp_features = construct_feature_mask(
+            feature_mask, formatted_inputs
+        )
+        num_features_list = torch.arange(num_interp_features, dtype=torch.float)
+        denom = num_features_list * (num_interp_features - num_features_list)
+        probs = (num_interp_features - 1) / denom
+        probs[0] = 0.0
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/kernel_shap.py:277]: This replaces the inf where the denominator is 0.

File: captum/attr/_core/kernel_shap.py [hunk 12]
```diff
@@ -304,4 +284,48 @@ def attribute(  # type: ignore
...
+    def kernel_shap_similarity_kernel(
+        self, _, __, interpretable_sample: Tensor, **kwargs
+    ) -> Tensor:
+        assert (
+            "num_interp_features" in kwargs
+        ), "Must provide num_interp_features to use default similarity kernel"
+        num_selected_features = int(interpretable_sample.sum(dim=1).item())
+        num_features = kwargs["num_interp_features"]
+        if num_selected_features == 0 or num_selected_features == num_features:
+            # weight should be theoretically infinite when denom = 0
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/kernel_shap.py:299]: Good point, will make this comment clearer!

File: tests/attr/test_kernel_shap.py (hunk 6 above)
Comment (vivekmig) [reply tests/attr/test_kernel_shap.py:200]: This is increased from the default to ensure convergence within the necessary delta, random behavior is slightly different with the new sampling, which likely caused this.

File: captum/attr/_core/kernel_shap.py [hunk 13]
```diff
@@ -304,4 +284,48 @@ def attribute(  # type: ignore
...
+        num_features = kwargs["num_interp_features"]
+        yield torch.ones(1, num_features, device=device, dtype=torch.long)
+        yield torch.zeros(1, num_features, device=device, dtype=torch.long)
+        while True:
+            num_selected_features = kwargs["num_select_distribution"].sample()
+            rand_vals = torch.randn(1, num_features)
+            threshold = torch.kthvalue(
+                rand_vals, num_features - num_selected_features
+            ).values.item()
+            yield (rand_vals > threshold).to(device=device).long()
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/kernel_shap.py:331]: Sure, will add more documentation on this. There are a few other helper methods in Lime without type hints, so will add them together in a separate PR.

**Comment by vivekmig** [comment]
Thanks for the review @NarineK ! Addressed comments.

File: captum/attr/_core/kernel_shap.py [hunk 14]
```diff
@@ -304,4 +284,61 @@ def attribute(  # type: ignore
...
+        return torch.tensor([similarities])
+
+    def kernel_shap_perturb_generator(
+        self, original_inp: Union[Tensor, Tuple[Tensor, ...]], **kwargs
+    ) -> Generator[Tensor, None, None]:
+        r"""
+        Perturbations are sampled by the following process:
+         - Choose k (number of selected features), based on the distribution
+                p(k) = (M - 1) / (k * (M - k))
+            where M is the total number of features
```
Comment (NarineK) [inline captum/attr/_core/kernel_shap.py:317]: nit: total number of features in the interpretable space ?

**Review by NarineK — APPROVED** [review]
Thank you for the explanation! Looks great! Maybe you could add in the description that each of the (M choose k) samples has equal prob of getting chosen thus we do this:
```
rand_vals = torch.randn(1, num_features)
            threshold = torch.kthvalue( ... 
```
If I remember your explanation correctly.
